Server

leodido opened the pull request.

Commit:
New: Repositories are now able to list a copy of the elements they contain

Commit:
Update: Service provider config no more a Resource

Commit:
Update: Service provider config validations (refs #37)

samechelon committed:
Update: Testing of List method for SchemaRepository and ResourceTypeRepository

Commit:
New: Default server config

Commit:
New: Configuration phase for server

alelb committed:
Update: Testing call config() returns ServiceProviderConfig instance

alelb committed:
Fix: Correct directory path for resource_type repo

alelb committed:
Update: Testing repos have been created when calling config() func

Commit:
Update: Default variable for service provider config (scimd server)

Commit:
New: Setup of SCIMD server endpoints

Commit:
New: Middleware for not implemented (501) verbs

Commit:
Update: Insensitive check not implemented methods

alelb committed:
Update: Read schema and resource file from directory in config testing

Commit:
Update: Using helper library for utilities functions

Commit:
New: Authentication middleware + Bulk endpoint configurability

Commit:
Docs: Endpoint documentation

Commit:
New: Bind additional parameters in queries (list, get, searches)

Commit:
Update: Defaults support for Search structs

leogr commented on schemas/core/generic_repository_test.go:
```diff
@@ -81,3 +81,43 @@ func TestResourceTypeRepository(t *testing.T) {
 	// (todo): test lock
 }
+
+// (todo): test list
```
It was tested below, isn't it?

leodido replied:
Yeah, I'll remove this comment.

leogr commented on server/server.go:
```diff
@@ -0,0 +1,41 @@
+	switch authenticationType {
+	case strings.ToLower(HTTPBasic.String()):
+		return gin.BasicAuth(gin.Accounts{
+			"admin": "admin",
```
Pls, fetch credentials from config or add a `(todo)` just to do not forget

leodido replied:
Adding a TODO. Thanks.

leogr reviewed (approved):
It's ok for me, just some comments may be needed (see below)

Commit:
Update: Reorganizing logic of handlers

Commit:
Update: Refactor event dispatching

alelb committed:
Update: Server uses Storage middleware now

alelb committed:
Update: Server uses Set middleware to inject repos within the context now

Commit:
New: SCIM server handlers decoration

Commit:
Update: SCIMD command now uses decoration to create routes

Commit:
New: Service for static resources

Commit:
Update: Using decorators for static resources

alelb committed:
Fix: Abort request when no storage is found

Commit:
Fix: Unmarshalling SearchRequest ensures its schema field is not lost anymore

Commit:
Fix: Projection test requires schema is present now

Commit:
WIP: Query resources have concurrency issues

Commit:
Fix: Switching to the new version of mgo

leogr committed:
Update: Internal refactor of adapter's Get()

samechelon committed:
Update: Added testing case for Messages search

leogr committed:
Update: Attributes with sub-attributes will not be returned anymore by projection

cc @leodido
The result of projection is usually converted to a map prior to be passed to the database and there's no way to make map's items order predictable, thus path to a complex attribute and paths to its sub-attributes could be returned in any order.
Within mongo, different projection's items ordering will produce different query results (in particular when the parent attribute is listed after its sub-attributes), so the final behavior is unpredictable because map's items ordering could be different each time.
Currently, to mitigate the "parent and children ordering" issue we decided to return only leaf-attributes that's not affect the final projection.
Also, with this strategy, it's not possibile to exclude all sub-attributes by just excluding their parent attribute (because it's always ignored) so it makes impossibile to implement the behavior described by #51

leodido commented:
Needs a way to specify **configs** affecting the service command, but not in the service provider config.

Eg.,
- [x] Whether to enable the self resource or not
- [x] Defaults values (min, max) for count parameter (refs #55)

We also must implement one (or more) way to provision configuration.
Better if using a layer capable of abstracting the type of configuration source.

Eg.,
- [x] YAML file

_Please list here other things, editing this message_. 

cc @alelb, @leogr, @samechelon.

leodido commented:
This message is about **choices**.

Choices we have to do.

- [ ] How to handle the authenticated subject alias (eg., `/Me` endpoint) - Section 3.11 @ RFC7644
    - [ ] Permanent redirect (308) to `/User`
    - [ ] Process directly with `Location` header pointing to the aliased resource (ie., `User`)
    - [ ] Both way but configurable
- [x] How to handle the `/ServiceProviderConfigs` endpoint: `ServiceProviderConfig` is not represented as a resource, has not an identifier, a "get all" method does not make much sense
     - [x] <strike>Leave as is currently</strike>
     - [x] <strike>Use the `StaticResourceService`</strike>
     - [x] Create a NotIdentifiableResourceService
- [x] How to behave when storage server is not reachable
    - [x] @alelb can you confirm this is ok now?
- [x] How to enforce `meta.resourceType` to be present also when `attributes` (inclusions) is specified (refs #55)
    - [x] Make possible that parent and children have different values for same characteristic
    - [x] Make filtering paths working also on leafs
- [x] How to handle ambiguous attributes
    - [x] Only match against the base schema
- [x] What to do when user specifies an unknown attribute or excluded attribute
   - [x] <strike>Panicking (current implementation)</strike>
   - [x] Ignoring it

_Please list here other choices, editing this message_.

cc @alelb, @leogr, @samechelon.

Commit:
Update: Alias for authenticated subject via decorator

Commit:
Fix: Enforcing meta.resourceType attributes to be always returned (fixes #55)

leodido commented:
The subattributes `meta.resourceType` is enforced to be **always** returned.
Thus it is now an element of the minimal set of attributes returned.

This has required also some bug fixing, in particular to the `Contexts()` function that now correctly executes filtering also on subattributes which parent has already been filtered out.

cc @leogr

Commit:
New: Ability to check whether a path refers to a subattribute or not

Commit:
Fix: Included and excluded attributes are always contextualized now

Commit:
Fix: Remove strong assumptions on meta sub-attributes

samechelon committed:
Update: Added projection test cases

Commit:
Fix: Core resource can not contain any value of its extensions

For example this happens when a projection does not specify any attribute belonging to a schema extension.

Commit:
New: Initial implementation for the GET resource

Commit:
Fix: Panic when projection method gets unknown attrs

We could also ignoring them completely (preferably, imho).

leodido reviewed (commented):
Better ignoring unknown attributes imho.

What do you think about @leogr ?

Commit:
Update: Test about projecting subattributes

leogr commented:
@leodido 
As per RFC, service provider SHALL ignore extraneous attributes for interoperability purpose because they may exists in other schemas implemented by 3rd-party but not by the current service provider, thus I suggest to just filter out those attributes which cannot be contextualised (ie. `Projection` should not return them).

Also, I think we have to do not check for `Undefined()` paths here, because contextualised path cannot be undefined (if that was possible, it would mean that schema was invalid and it's not a `Projection` responsibility ensuring that).

Finally, I think also that this kind of API should never panic.

Commit:
Fix: Context() have to return nil when attribute is unknown

Commit:
Fix: Correctly handle the inclusion of all unknown attributes

Commit:
Fix: Marshal resource only when it has values (data)

leodido commented:
I finally fix'd (almost) everything we need to fix to make the `GET /<resource>/:id` to work.

Anyway there are already **some gotchas we need to discuss**.

## A. Inclusions

### 1. Existing attribute(s)

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?attributes=displayname,name.givenname`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "resourceType": "User"
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "displayName": "Tiffany Fork",
        "name": {
            "givenName": "Tiffany"
        }
    }
}
```

<strike>Notice the values of `meta.location`, `meta.created` and so on (ie., zero values since not grabbed from storage) ...

Imho we should fix this ... cc @leogr, @alelb </strike>

### 2. All unknown attributes

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?attributes=unk,ciao,pippo`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "location": "https://example.com/v2/Users/2819c223-7f76-453a-919d-ab1234567890",
        "resourceType": "User",
        "created": "2010-01-23T04:56:22Z",
        "lastModified": "2011-05-13T04:42:34Z",
        "version": "W/\"a330bc54f0671c9\""
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "active": true,
        "addresses": [
            {
                "...": "...",
            },
            {
                "...": "...",
            }
        ],
        "displayName": "Tiffany Fork",
        "...": "..."
}
```

<strike>Notice the strange (null) value of `lastModified` ... cc @alelb</strike>

### 3. No inclusions

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891`

JSON as (A.2), (B.2).

### 4. Existing extension attribute(s)

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?attributes=urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:costCenter,urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:manager.$ref,displayname`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "resourceType": "User"
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "displayName": "Tiffany Fork"
    },
    "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
        "costCenter": "4130",
        "manager": {
            "$ref": "../Users/26118915-6090-4610-87e4-49d8ca9f808d"
        }
    }
}
```

### 5. Existing reference attribute(s)

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?attributes=displayName,groups.$ref`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "resourceType": "User"
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "displayName": "Tiffany Fork",
        "groups": [
            {
                "$ref": "https://example.com/v2/Groups/fc348aa8-3835-40eb-a20b-c726e15c55b5"
            }
        ]
    }
}
```

## B. Exclusions

### 1. Existing attribute(s)

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?excludedAttributes=displayName,locale,nickName,profileurl,title,timezone,username,usertype,preferredlanguage,active,x509certificates.value,photos.value,photos.type,phonenumbers.type,phonenumbers.value,ims.type,ims.value,name.familyname,name.formatted,name.honorificsuffix,name.honorificprefix,emails.type,emails.value,groups.value,groups.display,addresses.type,addresses.streetaddress,addresses.type,addresses.region,addresses.country,addresses.formatted,addresses.postalcode,addresses.locality`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "location": "https://example.com/v2/Users/2819c223-7f76-453a-919d-ab1234567890",
        "resourceType": "User",
        "created": "2010-01-23T04:56:22Z",
        "lastModified": "2011-05-13T04:42:34Z",
        "version": "W/\"a330bc54f0671c9\""
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "addresses": [
            {
                "primary": true
            },
            {}
        ],
        "emails": [
            {
                "primary": true
            },
            {}
        ],
        "groups": [
            {
                "$ref": "https://example.com/v2/Groups/fc348aa8-3835-40eb-a20b-c726e15c55b5"
            }
        ],
        "ims": [
            {}
        ],
        "name": {
            "givenName": "Tiffany",
            "middleName": "Geraldine"
        },
        "phoneNumbers": [
            {},
            {}
        ],
        "photos": [
            {},
            {}
        ],
        "x509Certificates": [
            {}
        ]
    },
    "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
        "costCenter": "4130",
        "department": "Tour Operations",
        "division": "Theme Park",
        "employeeNumber": "701984",
        "manager": {
            "$ref": "../Users/26118915-6090-4610-87e4-49d8ca9f808d",
            "displayName": "John Smith",
            "value": "26118915-6090-4610-87e4-49d8ca9f808d"
        },
        "organization": "Universal Studios"
    }
}
```

Notice the complex attributes with their zero values (eg. `[], {}`) ...

Similar to the case A.1.

We could think to a layer responsible to eliminate zero values from responses. cc @leogr, @alelb 

And also pay attention to do not eliminate attributes for which the `nil` value is possible.

### 2. All unknown exclusions

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?excludeAttributes=pippo,pluto`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "location": "https://example.com/v2/Users/2819c223-7f76-453a-919d-ab1234567890",
        "resourceType": "User",
        "created": "2010-01-23T04:56:22Z",
        "lastModified": "2011-05-13T04:42:34Z",
        "version": "W/\"a330bc54f0671c9\""
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "active": true,
        "...": "..."
    },
    "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
        "costCenter": "4130",
        "...": "..."
    }
}
```

### 3. No exclusions

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891`

JSON as (A.2), (B.2).

### 4. Existing extension attribute(s)

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?excludedAttributes=urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:costcenter,urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:department`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "location": "https://example.com/v2/Users/2819c223-7f76-453a-919d-ab1234567890",
        "resourceType": "User",
        "created": "2010-01-23T04:56:22Z",
        "lastModified": "2011-05-13T04:42:34Z",
        "version": "W/\"a330bc54f0671c9\""
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
   "urn:ietf:params:scim:schemas:core:2.0:User": {
        "active": true,
        "...": "..."
    },
    "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
        "division": "Theme Park",
        "employeeNumber": "701984",
        "manager": {
            "$ref": "../Users/26118915-6090-4610-87e4-49d8ca9f808d",
            "displayName": "John Smith",
            "value": "26118915-6090-4610-87e4-49d8ca9f808d"
        },
        "organization": "Universal Studios"
    }
}
```
### 5. Existing reference attribute(s)

`GET localhost:8787/v2/Users/2819c223-7f76-453a-919d-ab1234567891?excludedAttributes=urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:manager.value,urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:manager.$ref,groups.$ref`

```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "location": "https://example.com/v2/Users/2819c223-7f76-453a-919d-ab1234567890",
        "resourceType": "User",
        "created": "2010-01-23T04:56:22Z",
        "lastModified": "2011-05-13T04:42:34Z",
        "version": "W/\"a330bc54f0671c9\""
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
   "urn:ietf:params:scim:schemas:core:2.0:User": {
        "active": true,
        "...": "...",
        "groups": [
            {
                "display": "Employees",
                "value": "fc348aa8-3835-40eb-a20b-c726e15c55b5"
            }
        ],
        "...": "..."
    },
    "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
        "costCenter": "4130",
        "department": "Tour Operations",
        "division": "Theme Park",
        "employeeNumber": "701984",
        "manager": {
            "displayName": "John Smith"
        },
        "organization": "Universal Studios"
    }
}
```

## C. Both

_TBD_

alelb committed:
New: First implementation of the POST resource

samechelon committed:
Update: Unknown attributes testing cases

Commit:
New: Support for reference (sub) attributes

leogr commented:
For **inclusion**
With the only exception of `meta.resourceType` that must be always present in our implementation, there're two things you have to check:
- there's a typo (`lastMod` instead of `lastModification`) [here](https://github.com/fabbricadigitale/scimd/blob/6451191415bd9e4683a5e6ede4b08077d1c97127/storage/mongo/adapter.go#L271)
- and also note:
```go
// Meta ...
type Meta struct {
	Location     string     `json:"location" validate:"uri,required"`
	ResourceType string     `json:"resourceType" validate:"required"`
	Created      *time.Time `json:"created" validate:"required"`
	LastModified *time.Time `json:"lastModified" validate:"required"`
	Version      string     `json:"version,omitempty"`
}
```
almost all are `required` and haven't `omitempty`, but should they be so?
Please, check the RFC.

That said I didn't get why `externalId` is returned, even it has `omitempty`

leogr commented:
For **exclusion**
About complex attributes with their zero values (eg. [], {}), I think it's an issue with mongo projection. I'm not sure, but it may happen when you project a sub-attribute that's not present within the document.
Anywau, something like that:
```json
        "phoneNumbers": [
            {},
            {}
        ],
```
it's not `nil` nor empty, so cannot be a JSON marshal responsibility to clean up them, event they have `omitempty`.

alelb committed:
New: First implementation for the put resource

Commit:
Fix: Last modified meta field works now

leodido commented:
About your first comment you were right.

I fixed the typo and setup the other fields of meta as **not** `required` and `omitempty`.

About your last comment investigating I found the following.

Assume a query with `excludedAttributes=emails.type,emails.value`.

The object returned from the mongo adapter will contain:

```
emails:[map[primary:true] map[]]
``` 

Notice the second field (ie., the empty map).

And then I found that resource.Resource MarshalJSON impl. does [not implement](https://github.com/fabbricadigitale/scimd/blob/feature/server/schemas/resource/resource.go#L112) `omitempty` currently.

Commit:
Fix: External id omit when empty now works

leogr commented:
@leodido note that “externalId" is actually missing from `toResource` [here](https://github.com/fabbricadigitale/scimd/blob/feature/server/storage/mongo/adapter.go#L199).
Also, it should be better if the "omitempty" check was applied to `toDoc` too (just to avoid saving the empty field, not mandatory).

leogr commented:
About
```
emails:[map[primary:true] map[]]
```
IMHO, the storage should not return that, however I'm not sure if we can fix that within the storage layer.

leogr reviewed (commented):
👍

alelb committed:
Update: Add decorators to resource field for validation when posting a resource

samechelon committed:
New: Testing Contexts() method (refs #57)

alelb committed:
New: First implementation of delete resource api/handler

leogr commented:
@leodido I noticed that response examples above are wrong.
For instance the following:
```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "resourceType": "User"
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "urn:ietf:params:scim:schemas:core:2.0:User": {
        "displayName": "Tiffany Fork",
        "name": {
            "givenName": "Tiffany"
        }
    }
}
``` 
should be so:
```json
{
    "id": "2819c223-7f76-453a-919d-ab1234567891",
    "meta": {
        "resourceType": "User"
    },
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User"
    ],
    "displayName": "Tiffany Fork",
    "name": {
         "givenName": "Tiffany"
     }
}
```
Because only extensions' attributes must be enclosed within base schema's attributes must not be enclosed per namespace within a complex, base schema attributes must not.

See https://tools.ietf.org/html/rfc7644#section-3.4.1

Marshalling/Unmarshalling should be implemented according with those rules.

What's wrong?

leodido commented:
@leogr you're right.

Probably the cause is this [line](https://github.com/fabbricadigitale/scimd/blob/feature/server/schemas/resource/resource.go#L135).

leogr commented:
Right, the bug is there. I'll try to fix it tomorrow morning. 👍

leogr committed:
Fix: Marshal schema attrs to the top level

alelb committed:
Update: Add logic to retrieve the resource created/updated

leogr committed:
Fix: DataType casting panics with zero value

leogr committed:
Update: Resource fetched from storage must have a minimal attribute set always

alelb committed:
Update: Add ResourceNotFoundError to wrap mongo resource not found error

leogr committed:
Fix: Storage should not return empty objects

leogr commented:
Hey you, by https://github.com/fabbricadigitale/scimd/pull/46/commits/166b561e2e564149be87de9467e3acf3217d38f0 I've fixed empty `{}` issue. Double check please, and let me know if it works properly.

leodido commented on server/resource.go:
```diff
@@ -156,10 +159,13 @@ func (rs *ResourceService) Put(c *gin.Context) {
+	res, err := update.Resource(store.(storage.Storer), rs.rt, id, contents)
+	if err != nil {
+		log.Println("(todo) > handle error")
+		c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
```
We need to throw our own scim errors.

leodido reviewed (commented):
Ok =)

samechelon committed:
Update: Added test case for Paths where fx is nil

samechelon committed:
Update: Mold's max tests done using testing table

alelb committed:
New: ValidateRequired api first implementation

alelb committed:
Update: ValidateRequired implementation for creating and updating resources

alelb commented:
@leodido , @leogr  Take a look to my [last commit](46/commits/58c6233d870e61c3e23bd952f14edf3b1c81aee5) please.
I moved ValidateRequired in `required` package to avoid import cycle with `.../api/attr` package.
We can move all functions to perform validation in a `shared` package. What do you think about this?
Any other proposals? 
Thanks!

leodido commented:
@alelb regarding validations, as discussed yersterday.

We could profit from gin validation capabilities for the common attributes struct.

Rather for the data map (complex) of the resource we could write some line of code that will perform structural validation based on the attribute characteristics of the attribute (ie., keys) contained in the data map.

mmenonifd committed:
New: Password hashing (refs #48)

leodido commented:
Hey @marcomenoni welcome!

Thanks a lot for the contribution. :trophy: 

Only one main observation.

We should move away the logic (password hashing via bcrypt) implemented by your listeners into a dedicated package.
So to split the adapter from its listeners: the idea is that they have to be plug-n-play (imagine another way of hashing password implemented via another listener handler) and also storage agnostic.

So i propose to move this part of the code in another package.
I propose to call it listeners, within storage/listeners folder.

This way it could be plugged one day also on other kind of storage without duplicating code for example. And the code is well separated.

marcomenoni commented:
ok, I am going to move the listeners in a dedicated package

alelb committed:
New: Search handler implementation + error handling

mmenonifd committed:
Update: Storage listener handlers have their own dedicated package now

leodido commented on storage/mongo/adapter.go:
```diff
@@ -53,7 +52,7 @@ func New(url, db, collection string) (storage.Storer, error) {
 	adapter.Dispatcher = event.NewDispatcher(0)
 	adapter.Emitter().Use("*", emitter.Void)
-	adapter.addListeners()
+	listeners.AddListeners(adapter.Emitter())
```
This line should be not here to avoid coupling as said.

We should attach/plug listeners within an upper tier (eg., the command, a server middleware or similar).

```go
...
ada, err := mongo.New(...)
listeners.AddListeners(ada.Emitter())
...
```

alelb committed:
Update: Error handling improvements

leodido committed:
Update: Moving hooks to storage middleware

alelb committed:
Fix: Attributes projection is also performed when there aren't attributes specified (fixes #64)

leogr reviewed (commented):
👍 That's correct.

cc @leodido @alelb 
Also mind that projection must be related to resource types and schema loaded within the service provider (ie. attribute stored in the db but not configured must be not returned).
This fix works in this way, because `resTypes` should reflect the current config.

leodido reviewed (commented):
Awesome, thanks @alelb but please commit regression tests.

leodido committed:
New: Setup harness tests

leodido committed:
New: Harness tests (API)

alelb committed:
Update: Projection regression test

alelb committed:
Update: Testing Attributes function for projection

alelb committed:
Fix: Typo in CommonAttributes tag

Co-authored-by: Leonardo Di Donato <leodidonato@gmail.com>

Commit:
Update: Ability to load and store elem within repositories from bytes

Plus a huge amount of fixes within tests ...

Commit:
New: Configurability

SCIMD gains its first configuration awareness.

Configurable:
- storage parameters
- directory where default things are (service provider config, resource types, schemas)

How:
- Via ENV variables.
- Through a .scimd.yml file in the same directory where the daemon is

samechelon committed:
Update: Replaced Get with PushFromFile for ResourceTypeRepository and SchemaRepository

samechelon committed:
Update: Added test cases and replaced Get with Pull

samechelon committed:
Update: Testing of PushFromData for SchemaRepository and ResourceTypeRepository

samechelon committed:
New: Config default's Group ResourceType initialization

samechelon committed:
New: Config default's User ResourceType initialization

leogr reviewed (commented):
Why static configs? Are JSONs not enough?

cc @leodido

leodido commented:
JSON file needs to be embedded within the binary (as I done last week) but this results in a lot of extra space and computation.

Furthermore loading from paths (substituted with virtual ones pointing to the embedded representation of JSON files, at runtime) over complicates the configuration system.

Commit:
New: Defaults

Commit:
Update: Direct push of elements within repositories

Commit:
Update: Configuration phase uses new golang defaults now

leodido commented on schemas/core/generic_repository.go:
```diff
@@ -55,15 +56,8 @@ func (repo *repositoryGeneric) Pull(key string) *Elem {
-		return elem, err
-	}
-
+// PushFromData allows to store an elem within its repository
```
Do not forget to fix typo.

leodido committed:
New: Configuration for pagination page size (fixes #55)

leodido committed:
New: Configuration validators

leodido committed:
Update: Configuration validators

leodido committed:
New: Configuration field for self resource

samechelon committed:
Fix: Comment typo wrong issue number

samechelon committed:
Update: Added validation to Group and User defaults

Commit:
New: Custom validator to check whether path exists

samechelon committed:
Update: Testing Push

Commit:
New: Custom validators for directories and files

samechelon committed:
New: Testing pathExists validator

samechelon committed:
Update: Added testing cases for pathExists

samechelon committed:
New: Testing isDir validator

Commit:
New: CLI setup + version command

Version string can be injected at build time.

Eg.,
go build -ldflags "-X github.com/fabbricadigitale/scimd/cmd.version=`git rev-parse HEAD`"

samechelon committed:
New: Testing isFile validator

Commit:
New: CLI

alelb committed:
Fix: Typo in testdata/list

alelb committed:
Fix: ScimError implements error() correctly

alelb committed:
Update: Explicit error handling for Projection(), Paths() and Contexts()

samechelon committed:
New: Added Clean() for repositories

samechelon committed:
Update: Testing Clean

alelb committed:
Fix: Typo in toResource()

alelb committed:
Update: Add omitempty to Location validator

alelb committed:
Fix: Integration Query API tests now work

Commit:
New: I18n for validator errors

Commit:
Update: Export validation logic for checks on paths and directories

alelb committed:
Update: Add more test cases for Normalize() testing

Commit:
New: Better management of validation errors for configurations

Commit:
New: CLI support for flags (bound to configs)

Commit:
New: Commands to let users extract the default configurations

alelb committed:
Fix: Add omitempty in SortOrder and StartIndex tag

alelb committed:
Update: Add check for resourceTypes parameter

Commit:
Update: Service provider configuration access

Commit:
New: Lookup configuration file also in home directory

Commit:
New: Print config CLI command

alelb committed:
New: Static resource endpoint implementations

Commit:
New: CLI command to obtain service provider config

alelb committed:
Update: Error message when storage is unreachable

Commit:
Fix: Wrong default endpoint for resource types

Commit:
Fix: Static service for service provider config

samechelon committed:
Update: Testing of toResource

Commit:
Update: Flags

Commit:
New: Harness tests for static resource services

GET /v2/Schemas/<id>
GET /v2/ResourceTypes/<id>
GET /v2/Schemas (missing check, implementation not complete)
GET /v2/ResourceTypes (missing check, implementation not complete)

Commit:
New: Harness test for service provider config service

leodido commented:
We are almost there guys ...

cc @fabbricadigitale/openinnovation

Commit:
New: Command to download default resource types and schemas

Commit:
Update: Testing exclusions of 1st level attribute with or without subattributes

alelb committed:
Update: Change list.resources type to slice of interface

alelb committed:
Update: Wrap static resources in ListResponse struct

alelb committed:
Fix: ConvertToMongo without filter

Co-authored-by: Leonardo Di Donato <leodidonato@gmail.com>

Commit:
Update: Enable or disable self resource by config value

Commit:
Fix: Copy resource type to avoid service provides always the last one

Co-authored-by: Alessandro La Bella <sandrone.lb@gmail.com>

samechelon committed:
Update: Harness test getting user
